src/behaviour.rs

```diff
@@ -211,12 +200,101 @@ impl<P: StoreParams> Bitswap<P> {
     }
 }
 
+impl<P: StoreParams> Bitswap<P> {
+    /// Processes an incoming bitswap request.
+    fn inject_request(&mut self, channel: BitswapChannel, request: BitswapRequest) {
+        let response = match request.ty {
+            RequestType::Have => {
+                let have = self.store.contains(&request.cid).ok().unwrap_or_default();
+                if have {
+                    RESPONSES_TOTAL.with_label_values(&["have"]).inc();
+                } else {
+                    RESPONSES_TOTAL.with_label_values(&["dont_have"]).inc();
+                }
+                tracing::trace!("have {}", have);
+                BitswapResponse::Have(have)
+            }
+            RequestType::Block => {
+                let block = self.store.get(&request.cid).ok().unwrap_or_default();
+                if let Some(data) = block {
+                    RESPONSES_TOTAL.with_label_values(&["block"]).inc();
+                    SENT_BLOCK_BYTES.inc_by(data.len() as u64);
+                    tracing::trace!("block {}", data.len());
+                    BitswapResponse::Block(data)
+                } else {
+                    RESPONSES_TOTAL.with_label_values(&["dont_have"]).inc();
+                    tracing::trace!("have false");
+                    BitswapResponse::Have(false)
+                }
+            }
+        };
+        match channel {
+            BitswapChannel::Bitswap(channel) => {
+                self.inner.send_response(channel, response).ok();
+            }
+            #[cfg(feature = "compat")]
+            BitswapChannel::Compat(peer_id) => {
+                let msg = CompatMessage::Response(request.cid, response);
+                self.compat.push_back((peer_id, msg));
+            }
+        }
+    }
+
+    /// Processes an incoming bitswap response.
+    fn inject_response(
+        &mut self,
+        id: BitswapId,
+        peer: PeerId,
+        response: BitswapResponse,
+    ) -> Option<BitswapEvent> {
+        if let Some(id) = self.requests.remove(&id) {
```

Comment:
Is it likely that there will be multiple requests in this? If so, you could do the following: grab them all, and then answer the queries in bulk. E.g. collect all `BitswapResponse::Block` and then do something like `store.insert_many()`.

I know this does not look elegant, but it is the only way to get good perf out of databases that flush/sync after write.

The alternative of just letting the db flush every 500ms or whatever is not a good solution if you are paranoid about your data. And if you are not paranoid about your data, why are you using all this hashing etc. in the first place :-)

Reply:
I guess I'll leave the batching, avoid flushing, as a follow up.